projects/js-packages/react-data-sync-client/src/DataSyncHooks.ts

```diff
@@ -125,3 +126,118 @@ export function useDataSync<
 		useMutation( { ...mutationConfigDefaults, ...config.mutation } ),
 	];
 }
+
+/**
+ * Use React Query mutations to dispatch custom DataSync Actions.
+ *
+ * ### Usage:
+ *
+ * ```ts
+ * const action = useDataSyncAction<{ foo: string }>()( 'namespace', 'key', 'action', schema, schema, callback );
+ * ``` Notice double function call ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+ *
+ * ## Anatomy of a callback:
+ * Return value to set the new state.
+ * Throw an error to cancel.
+ * Return undefined to keep the current state.
+ * ```ts
+ * callback: ( result: ActionResult, currentValue: State ) => void | State,
+ * ```
+ *
+ * ### Trigger a mutation:
+ * ```ts
+ * action.mutate( { foo: 'bar' } );
+ * ````
+ *
+ *
+ */
+export function useDataSyncAction< ActionData extends RequestParams >() {
+	/*
+	 * Yes, this is a function that returns a function.
+	 * Also yes - this is dumb.
+	 * But TypeScript still doesn't allow partial type inference.
+	 *
+	 * This is what happens when you rush development
+	 * - you get legacy code that can't be fixed for years.
+	 * @see https://github.com/microsoft/TypeScript/issues/26242
+	 */
+
+	return function <
+		Schema extends z.ZodSchema,
+		State extends z.infer< Schema >,
+		Key extends string,
+		ActionName extends string,
+		ActionSchema extends z.ZodSchema,
+		ActionResult extends z.infer< ActionSchema >,
+		Config extends DataSyncConfig< ActionSchema, ActionData >[ 'mutation' ],
+	>(
+		namespace: string,
+		key: Key,
+		name: ActionName,
+		stateSchema: Schema,
+		actionSchema: ActionSchema,
+		callback: ( result: ActionResult, currentValue: State ) => void | State,
+		config: Config = {} as Config,
+		params: Record< string, string | number > = {}
+	) {
```

Comment:
"No," I said to myself. "Surely this isn't the only way to deal with complex typing issues like this," I said.
I fell down the rabbit hole of looking into this and 😱 it _is_ how you do it. I shall never be the same again.

Aside from this ugly corner, which I acknowledge is a workaround for TypeScript, this is looking like a great solution 👍